src/java.base/share/classes/java/util/random/RandomGenerator.java

```diff
@@ -634,9 +640,13 @@ default int nextInt() {
      *
      * @throws IllegalArgumentException if {@code bound} is not positive
      *
-     * @implSpec The default implementation simply calls
-     * {@link RandomSupport#checkBound checkBound}(bound) and then
-     * {@link RandomSupport#boundedNextInt boundedNextInt}(this, bound).
+     * @implSpec The default implementation checks that {@code bound} is a
+     * positive int. Then invokes {@code nextInt()}, limiting the result to
+     * be greater equal zero and less than {@code bound}. If {@code bound} is a
+     * power of two then limiting is a simple masking operation. Otherwise, a
+     * new result is re-calculated by averaging the previous result and
+     * {@code nextInt()} until the final result is greater equal zero and less
```

Comment:
I don't see how 'averaging' comes from the invocation of boundedNextInt which appears to choose the first candidate that meets the criteria.
The comment also applies to nextLong overloads.

Reply:
If the first value calculated doesn't meet the criteria, then you either mask (power of 2) or go into this loop with does the averaging. The intent is not to bias the outcome.

```
                for (long u = r >>> 1;            // ensure nonnegative
                     u + m - (r = u % n) < 0L;    // rejection check
                     u = rng.nextLong() >>> 1) // retry
                    ;
```

Reply:
Yea you're right. I was focused on the divide by two, which appears to be to just remove the sign. Clearer interpretation:

```
r = rng.nextLong();
n = range;
m = n - 1;                   // mask
for (long u = r / 2;         // ensure nonnegative
     u + m < r;              // rejection check
     u = rng.nextLong() / 2) // retry
    r = u % n;
```